### PR TITLE
Arbitrary logic gravity side support.

### PR DESCRIPTION
**Pull request by jackzampolin**
closes https://github.com/althea-net/cosmos-gravity-bridge/issues/189

cc @jkilpatr @jtremback

### CODE REVIEW

**Comment by Jehan** [comment]
> @Jehan

Why was I summoned here? You probably meant to notify someone else as I don't think I have a relationship with this project. Or do I miss anything?

**Comment by jackzampolin** [comment]
Oh @Jehan I'm sorry. I meant @jtremback 🤦‍♂️

**Comment by jtremback** [comment]
Finally... I meet the real Jehan

File: module/proto/peggy/v1/batch.proto [hunk 1]
```diff
@@ -23,3 +23,14 @@ message OutgoingTransferTx {
   ERC20Token erc20_token  = 4;
   ERC20Token erc20_fee    = 5;
 }
+
+// OutgoingLogicCall represents an individual logic call from Peggy to ETH
+message OutgoingLogicCall {
```
Comment (jtremback) [inline module/proto/peggy/v1/batch.proto:28]: Maybe I'm missing something here, but this doesn't seem to conform to the actual arguments for a logic call. Here they are for reference:

```
	struct LogicCallArgs {
		// Transfers out to the logic contract
		uint256[] transferAmounts;
		address[] transferTokenContracts;
		// The fees (transferred to msg.sender)
		uint256[] feeAmounts;
		address[] feeTokenContracts;
		// The arbitrary logic call
		address logicContractAddress;
		bytes payload;
		// Invalidation metadata
		uint256 timeOut;
		bytes32 invalidationId;
		uint256 invalidationNonce;
	}
```

First of all, a lot of them are named something different. Why?

I'm guessing "logic_nonce" is supposed to be the invalidation_nonce, and "call" is supposed to be the payload? I'm guessing "id" is supposed to be the invalidation_id? If so, it is the wrong type. It should be []bytes.

And we're totally missing "transfer_amounts". How can the e.g., the Uniswap liquidity redeeming logic redeem LP tokens if Gravity doesn't give it any LP tokens?

File: module/proto/peggy/v1/msgs.proto [hunk 2]
```diff
@@ -137,6 +140,24 @@ message MsgConfirmBatch {
 
 message MsgConfirmBatchResponse {}
 
+// MsgConfirmLogicCall
```
Comment (jtremback) [inline module/proto/peggy/v1/msgs.proto:143]: Comment seems to be for something else, not this code

File: module/proto/peggy/v1/batch.proto (hunk 1 above)
Comment (jtremback) [reply module/proto/peggy/v1/batch.proto:28]: I'll just write out what I think this should probably be.

```
message OutgoingLogicCall {
  repeated ERC20Token transfers     = 1;
  repeated ERC20Token fees     = 2;
  string logic_contract_address = 3;
  bytes payload = 4;
  uint64 timeout = 5;
  bytes invalidation_id = 6;
  unit64 invalidation_nonce = 7;
}
```

**Review by jtremback — CHANGES_REQUESTED** [review]
Looks ok from a skim, except for the OutgoingLogicCall proto message, which has serious issues

File: module/proto/peggy/v1/batch.proto (hunk 1 above)
Comment (jtremback) [reply module/proto/peggy/v1/batch.proto:28]: I will say that we might be able to name some of the arguments for the contract method something else. For example "invalidation_id" could definitely be changed to "invalidation_scope". If we decide to do that, I'll change it in the solidity as well.

**Comment by Jehan** [comment]
> Finally... I meet the real Jehan

Ahah I don't know if there are real or fake Jehan-s. That's my first name since birth, this is for sure though. 🙂
Anyway I'll leave you to your project everyone. No more off-topic, I guess! 👍

File: module/proto/peggy/v1/batch.proto (hunk 1 above)
Comment (jackzampolin) [reply module/proto/peggy/v1/batch.proto:28]: I'm fine with names as is. Will refactor to match here

File: module/proto/peggy/v1/msgs.proto (hunk 2 above)
Comment (jackzampolin) [reply module/proto/peggy/v1/msgs.proto:143]: copy pasta 🙈 =

**Pull request merged**